packages/react-i18n/src/index.jsx

```diff
@@ -0,0 +1,142 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	createContext,
+	useContext,
+	useEffect,
+	useMemo,
+	useReducer,
+} from '@wordpress/element';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { createHooks, defaultHooks } from '@wordpress/hooks';
+import { defaultI18n } from '@wordpress/i18n';
+
+/** @typedef {import('@wordpress/i18n').I18n} I18n */
+/** @typedef {import('@wordpress/hooks').Hooks} Hooks */
+
+const I18nContext = createContext(
+	makeContextValue( defaultI18n, defaultHooks )
+);
+
+export const I18nProvider = ( { children, i18n = defaultI18n } ) => {
+	const hooks = useMemo( createHooks, [] );
+
+	const [ update, forceUpdate ] = useReducer( () => [], [] );
+
+	// rerender translations whenever a hook is removed or added
+	useEffect( () => {
+		hooks.addAction( 'hookAdded', 'core/react-i18n/filters', () =>
+			forceUpdate()
+		);
+		hooks.addAction( 'hookRemoved', 'core/react-i18n/filters', () =>
+			forceUpdate()
+		);
+		return () => {
+			hooks.removeAction( 'hookAdded', 'core/react-i18n/filters' );
+			hooks.removeAction( 'hookRemoved', 'core/react-i18n/filters' );
+		};
+	}, [] );
```

Comment:
Should filtering perhaps better be handled by #27966?

Reply:
@leewillis77 is working on adding filters to the i18n package. I would be curious to learn how much those two PRs overlap.

Reply:
That's a question for @yuliyan, who added this feature in https://github.com/Automattic/wp-calypso/pull/44743, presumably to implement a i18n feature like Community Translator. @yuliyan did you implement the feature you needed the hooks for yet?

And would it be ok if these hooks were not in the `react-i18n` library, but directly in the `I18n` instance in `@wordpress/i18n`?

Another difference is that @leewillis77's work in #27966 and also the [Core `translate` PHP function](https://github.com/WordPress/WordPress/blob/master/wp-includes/l10n.php#L176-L205) only have hooks to modify the `gettext` _result_ after the translation, and not its arguments _before_ translation. @yuliyan's version has a `preTranslation` hook that can modify these arguments.

I'd personally prefer to move the hooks support to the core `I18n` class and simplify the React bindings as much as possible. But we need to make sure that the features that depend on these hooks are OK with that.

Reply:
> That's a question for @yuliyan, who added this feature in [Automattic/wp-calypso#44743](https://github.com/Automattic/wp-calypso/pull/44743), presumably to implement a i18n feature like Community Translator.

That's correct, but the implementations are still in a draft state and have not been released yet.



> And would it be ok if these hooks were not in the `react-i18n` library, but directly in the `I18n` instance in `@wordpress/i18n`?

I think it would be even better to have them directly in the `I18n` instance, though `react-i18n` bindings should still maintain reactivity when adding/removing hooks.



> Another difference is that @leewillis77's work in #27966 and also the [Core `translate` PHP function](https://github.com/WordPress/WordPress/blob/master/wp-includes/l10n.php#L176-L205) only have hooks to modify the `gettext` _result_ after the translation, and not its arguments _before_ translation. @yuliyan's version has a `preTranslation` hook that can modify these arguments.

I like the implementation in #27966 as it looks closer to the core PHP `gettext` function filters and essentially provides the same functionality as the `postTranslation` filter here.

As for the lack of `preTranslation` - I can't tell if it's a mandatory feature, but it would definitely come in handy. The first use case that pop into my head would be to use it for implementing a different translation lookup logic, in case the JED file uses translation keys other than the original string (e.g. hashes, ids, etc.).

Reply:
> I think it would be even better to have them directly in the `I18n` instance

Perfect! That means I can remove them from the `react-i18n` bindings in this PR.

> though `react-i18n` bindings should still maintain reactivity when adding/removing hooks.

That can be implemented in `i18n` by registering the `hookAdded`/`hookRemoved` callbacks and fire a `subscribe` event when a relevant `i18n.*` hook changes. At this moment, `subscribe` listeners are notified only on `setLocaleData()` calls, now they'll also be notified when hooks change.

> As for the lack of `preTranslation` - I can't tell if it's a mandatory feature, but it would definitely come in handy. The first use case that pop into my head would be to use it for implementing a different translation lookup logic, in case the JED file uses translation keys other than the original string

OK, but that sounds like a rather hypothetical use case to me, not something we need in near future. Or is it needed for any of the existing `i18n-calypso` hooks and filters that we use, e.g., for Community Translator or for highlighting untranslated strings in the UI?

The `pre_gettext` hooks can be added at any later time to `@wordpress/i18n`.